### ISSUE TITLE
Support subcommands

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/meow#69]

**Comment by muuvmuuv** [comment]
Just an idea but I thought about we could do it similar to [Listr](https://github.com/SamVerschueren/listr#usage)

```js
#!/usr/bin/env node
"use strict";
const meow = require("meow");
const foo = require(".");

const flags = {
  rainbow: {
    type: "boolean",
    alias: "r"
  }
}

const subCommands = {
  horseman: (flags /* parsed flags from parent */) => {
    return meow(
      `
      Usage
        $ horseman
      .....
    `,
      {
        name: {
          type: "string",
          default: "The Horseman"
        }
      }
    );
  }
}

const cli = meow(
  `
  Usage
    $ foo <input>

  Options
    --rainbow, -r  Include a rainbow

  Examples
    $ foo unicorns --rainbow
    🌈 unicorns 🌈
`,
  {
    flags,
    subCommands
  }
);

foo(cli.input[0], cli.flags);
```

**Comment by zzyyxxww** [comment]
I actually like that meow doesn't have subcommands. 
I'm not a fan of subcommands, my experience with cli parsers that implement them is usually awful, either because the API is too obscure or buggy, and can result in humongous user `cli.js` files (given a command tree is complex enough).

To have some subcommands in meow, you can do:
```javascript
import daemon from './cli-daemon'
import memory from './cli-memory'

const commands = { daemon, memory }
const cli = meow(...)
const c = cli.input.unshift()
if (!c) { cli.showHelp() }
commands[c](cli)
```
Then, every `./cli-<action>` module can decide to run meow again with some other flags while still retaining access to the global flags.

Now, there isn't a single monolithic `cli.js` file, every command and subcommand is encouraged to have its own file that will probably be easy to read. Even if you had a very complex tree of subsubsubcommands, this strategy would scale very well.

@sindresorhus maybe there should be a section in the readme describing a standard way of having subcommands.

**Comment by zzyyxxww** [comment]
I like meow because it's useful yet simple, but most implementations of subcommands are ridiculous, so I favor this change as long as user control is prioritized. 

I also think that the property should be called `commands` instead of `subcommands`, it's shorter and it's still correct. For example, git clone, git add and git push are all git commands.

Regarding @muuvmuuv's solution, I like that it doesn't support nesting of commands but it still has a couple of challenges that I think disqualifies it:
1) It doesn't tell you **which** command should it run.
2) It takes some control away by calling a function you provide it with. 

A more simple way is making `meow` to expect the `commands` property to be an `Array` of `String`,
It could even have suggestions in case you misspell a command (like git does).
```javascript
import fetch from './cli-fetch'
import sort from './cli-sort'

const options = { fetch, sort }
const { command, flags } = meow({
  commands: Object.keys(options),
  ...
})
options[command](flags)
// each cli-<action> module can use meow in turn too
```
```bash
$ foo michael
foo: 'michael' is not a foo command. See foo --help
The most similar commands are
    fetch
    sort
```
This way meow doesn't concern with calling functions and deciding which params to pass to these. 
The `Object.keys` usage could become a standard for deciding which command to use.

A problem with my solution is that sometimes commands are the second level, for example the `connect` in `docker network connect`. So the `commands` option should decide the index at which this should be.
```
// fictional-docker-swarm.js
meow({
  commands: {
    1: ['connect', 'create', 'disconnect', 'inspect', 'ls', 'prune', 'rm']
  }
})
```

It's simple but it requires a very specific format. This is open to criticism and I'm willing to push a PR in any case.

**Comment by MartinMuzatko** [comment]
I have written a tiny script to allow subcommands in this structure:
https://gist.github.com/MartinMuzatko/2922add0f84535a44f98a8e8fa729efa

**Comment by skoshx** [comment]
Any work being done on this feature? I like @muuvmuuv 's idea of having meow instances as subcommands, I think it's the most flexible, most simple & makes it easy to separate subcommands into different files for a cleaner and more maintainable codebase.

@sindresorhus Do you have any ideas/criticism to this approach? Was thinking that if no one is working on this that I could do it as soon as I have time.

**Comment by sindresorhus** [comment]
> Do you have any ideas/criticism to this approach?

Yeah, I'm fine with that approach, but it should be called `commands`.

**Comment by skoshx** [comment]
I agree that it should be called `commands`, I'll get on it as soon as possible, unless someone beats me to the punch :D

**Comment by aaronccasanova** [comment]
Hello 👋 I've been experimenting with an alternative approach for adding support for `subcommands` and would love to get folks thoughts and opinions on the API!

Here is my [WIP](https://github.com/sindresorhus/meow/pull/211)

**Comment by detj** [comment]
My attempt without changing anything in meow. Uses Dynamic Imports to load the additional command CLIs.

https://github.com/detj/meow-subcommands

I'm sure, there might be issues I'm unaware of. Feel free to offer suggestions.

**Comment by voxpelli** [comment]
I solved it like this in the @SocketDev CLI: https://github.com/SocketDev/socket-cli-js/blob/f45cbd408c0575dfe205cf968596f4506f15f44d/lib/utils/meow-with-subcommands.js#L33

Planned to extract it to its own module and might still do even though I'm no longer with Socket
